### PR TITLE
fix(tags): protect viewer's toggled tag from stale Nexus merges (#1781)

### LINKED ISSUE #1781: fix(tags): Nexus refetch reverts just-toggled tag
> **Status update (2026-05-11)**
> Implemented as a sessionStorage-based marker mechanism that survives F5 refresh, with no IndexedDB schema change. The long-term solution (multi-device drift) is still tracked in #1800. Earlier plans considered (IndexedDB marker, in-memory Map, SSE) are kept at the bottom of this issue for reference.

## Problem

Two separate problems, easy to confuse with each other.

### 1. Race: a stale Nexus response overwrites a fresh local mutation

`usePostTags.fetchInitialTags` fires a Nexus request on every mount. If the user toggles a tag while that request is in flight, the (now stale) response comes back and `LocalPostTagService.mergeTags` writes it into IndexedDB, undoing the local change.

Repro on `/home`:

1. A post is rendered with one of the viewer's tags highlighted.
2. Viewer clicks the tag to remove it. Local commit succeeds, UI updates correctly.
3. About one second later, the in-flight Nexus response from the page mount returns. Nexus has not indexed the homeserver delete yet, so it replies with `relationship: true` and viewer in `taggers`. `mergeTags` writes that into IndexedDB.
4. The tag is highlighted again.
5. If the user clicks again to remove it, the homeserver returns 404 because the tag is already deleted there.

QA team has flagged the visible flicker as a high-priority bug.

### 2. Drift: `mergeTags` cannot accept "the data went down"

Even when the race does not happen, once IndexedDB and Nexus disagree, the local cache is stuck. Refreshing later — even after Nexus has fully caught up — does not fix it.

Two lines in `src/core/services/local/tag/post/tag.post.ts` are the structural cause:

```ts
// line 175 — union of taggers
const mergedTaggers = [...new Set([...(existingTag.taggers ?? []), ...(newTag.taggers ?? [])])];

// line 180 — max of counts
taggers_count: Math.max(existingTag.taggers_count ?? 0, newTag.taggers_count ?? mergedTaggers.length),
```

- The **union** of taggers means any later Nexus response that still lists the viewer will re-add the viewer, even after the viewer removed themselves. (If a tagger appears on either side, they stay — removals never carry over.)
- `Math.max` on the count means the count never goes down, even when Nexus says it should.
- Also, `mergeTags` never deletes a label from IndexedDB if Nexus omits it (sole-tagger delete case — deferred to a follow-up PR).

Concrete example for the count problem:

```
Start:  tag "abc", taggers = [viewer], taggers_count = 1
Viewer removes themselves locally → local: taggers = [], taggers_count = 0
Stale Nexus response arrives:    taggers = [viewer], taggers_count = 1
mergeTags result:
  taggers       = union([], [viewer])      = [viewer]    ← back from the dead
  taggers_count = Math.max(0, 1)           = 1           ← back from the dead

Tens of seconds later, Nexus catches up. Refresh → fresh response:
  taggers = [], taggers_count = 0
But local IndexedDB now has [viewer] / 1 from the earlier merge.
mergeTags result:
  taggers       = union([viewer], [])      = [viewer]    ← stuck
  taggers_count = Math.max(1, 0)           = 1           ← stuck
```

Result: once drift starts, every refresh shows the wrong state, because `mergeTags` cannot write smaller numbers.

(Note: the `Math.max` in `usePostTags.tsx` for `loadedCountRef` is unrelated — that one is for pagination skip math.)

## Comparison of considered approaches

| Approach | Race + Drift Protection | F5 Mid-flicker | Multi-device | Schema/DB Cost | Effort | Notes |
|---|---|---|---|---|---|---|
| IndexedDB marker | Yes | Yes | No | New table + DB version bump | Medium | Persists, shared across tabs |
| In-memory Map | Yes | No | No | None | Very low | Marker lost on F5 |
| **sessionStorage marker (chosen)** | Yes | Yes | No | None | Low | Survives F5, scoped per tab |
| SSE (PR #1787 style) | Yes (with HS reads) | Yes | Yes | Persistent connection + viewer-tag index | Very high | Overkill for tag data — viewer tags don't need real-time push the way the mute list does |
| HS-direct fetch | Yes | Yes | Yes | New viewer-tag index table | Very high | Long-term plan, tracked in #1800 |

Why sessionStorage:

- Solves the visible flicker bug (the high-priority pain point).
- Survives F5 refresh — bigger win than the in-memory option.
- No IndexedDB schema change, no DB version bump.
- Minimal code surface.
- Multi-device drift is intentionally deferred to long-term (#1800).

## Short-term solution

Keep "fetch on every mount" — the post page should reflect new replies and other interactions on entry. Fix the race and the drift inside `mergeTags`, with a small sessionStorage marker to protect viewer-only fields.

### A. Viewer-mutation marker in sessionStorage

When the viewer toggles a tag, we store a marker so we can recognize "the viewer just changed this — ignore stale Nexus data for this field for the next 5 minutes."

Key format:

```
pubky-app:viewer-tag-marker:${currentUserPubky}:${postId}:${label}
```

Value: `JSON.stringify({ op, ts, expiresAt })` where `expiresAt = ts + 300_000` (5 minutes).

`op` reuses `HttpMethod`: `HttpMethod.PUT` (= viewer created the tag) or `HttpMethod.DELETE` (= viewer removed it). This matches the existing convention used by follow/mute (`commitFollow(eventType: HttpMethod, ...)`).

Written from `LocalPostTagService.create` and `LocalPostTagService.delete`. The label in the key is lowercased so reads match writes even if Nexus's response uses a different case.

The user-pubky in the key gives free isolation when several accounts share a device.

Marker entries are read through a single `parseMarker` helper that validates `op`, `ts`, and `expiresAt` together — anything that fails the shape check (corrupted JSON, unknown op, missing field) is dropped instead of treated as a marker.

### B. Redesigned `mergeTags` (per-field policy)

**Important context — Nexus `getPostTags` response is doubly limited:**

- The response paginates over **labels** (`skip` / `limit`). Labels not in the response may exist on later pages.
- Within each label, the `taggers` array is **truncated to the top-N taggers** (the full taggers list is served by a separate `getPostTaggers` endpoint). So `taggers` is a sample, not the full set.

This is why the original `mergeTags` was conservative: a flat "replace with Nexus" would drop locally-known taggers (including the viewer) that fall outside the top-N.

So the redesigned `mergeTags` uses a different policy per field:

| Field | Policy |
|---|---|
| `taggers_count` | Default: replace with the Nexus value (drop `Math.max`). Exception: if the marker overrides Nexus's view of viewer membership, adjust by ±1 so the count stays consistent with `taggers` / `relationship`. |
| `relationship` | If an active marker exists, follow the marker (`PUT` → true, `DELETE` → false). Otherwise follow Nexus. |
| Viewer entry in `taggers` | If an active marker exists, follow the marker. Otherwise derive from Nexus's `relationship`. |
| Other users in `taggers` | Keep the union of existing + Nexus, viewer excluded (the viewer's slot is decided above). Avoids dropping cached taggers that fall outside Nexus's truncated top-N. |
| Labels not in the response | Leave alone (pagination safety). Cleaning up "sole-tagger gone from Nexus" labels is deferred to a follow-up PR. |

This trade-off keeps the protections the original code was reaching for, and only changes the parts that drift.

### C. Marker lifecycle

- TTL: **5 minutes** (`MARKER_TTL_MS = 300_000`). After that the Nexus indexer has likely caught up, so the marker no longer adds protection value.
- Lazy delete on read: when `get` finds an expired or invalid marker, it removes the key as a side effect.
- Side-effect GC on every `mergeTags` call: `ViewerTagMarkerStorage.sweepExpired()` walks prefix-matching keys and drops expired or invalid ones so stale markers don't accumulate.
- Tab close → sessionStorage clears automatically.
- Logout → `AuthController.cleanupLocalState` calls `TagApplication.clearViewerMarkers(pubky)`, which clears all keys with the `pubky-app:viewer-tag-marker:${pubky}:` prefix for the current user.

### D. Homeserver 404 on DELETE = already deleted

`TagApplication.commitDelete`'s catch now treats a 404 from the homeserver as "the tag is already gone, end state matches the local delete" instead of rolling back. Before this fix, a delete against a ghost tag would 404 and the rollback would re-create the tag locally, leaving the user with a tag they could not remove. With the new behavior the local delete stands and the marker stays as `DELETE`. Other error codes still trigger the existing rollback.

### What this does NOT solve

- **Multi-device drift.** Markers are per-tab, per-device by design. Tracked in #1800.
- **Multi-tab on the same device.** A marker created in tab A does not protect tab B. Low frequency, accepted.
- **Sole-tagger ghost label cleanup** when another user removed their last-tagger tag. Local IndexedDB keeps the label until a follow-up PR adds the first-page missing-label cleanup.

### Why no AbortController

Cancelling the in-flight Nexus fetch would also lose other users' updates that came in the same response. With `mergeTags` redesigned to protect viewer-only fields, the response is safe to apply.

### Files changed

- `src/config/viewerTagMarker.ts` (new) — `MARKER_TTL_MS = 300_000`, `VIEWER_TAG_MARKER_STORAGE_PREFIX`.
- `src/core/services/local/tag/post/viewerTagMarkerStorage.ts` (new) — sessionStorage helper with `set` / `get` / `sweepExpired` / `clearForUser` and `parseMarker` shape validation.
- `src/core/services/local/tag/post/tag.post.ts` — write markers in `create`/`delete`, redesign `mergeTags`, call `sweepExpired` from `mergeTags`.
- `src/core/application/tag/tag.ts` — 404-on-DELETE handling in `commitDelete`, new `TagApplication.clearViewerMarkers` wrapper.
- `src/core/application/post/post.ts` — thread `viewerId` through `fetchTags` into `mergeTags`.
- `src/core/controllers/auth/auth.ts` — call `TagApplication.clearViewerMarkers` from `cleanupLocalState` before the auth store resets.

Tests added/extended:

- `src/core/services/local/tag/post/viewerTagMarkerStorage.test.ts` (new) — storage round-trip, label normalization, TTL expiry, parseMarker rejecting unknown `op` / wrong shapes / corrupted JSON, `sweepExpired`, `clearForUser`.
- `src/core/services/local/tag/post/tag.post.test.ts` — marker writes on `create`/`delete`, mergeTags per-field policy (count replace, ±1 adjustment under marker, union for other taggers, missing-label preservation), viewerId-null path, drift auto-correction, sweep wired into mergeTags.
- `src/core/application/tag/tag.test.ts` — 404-on-DELETE handling and `clearViewerMarkers` delegation.

## Long-term solution

Tracked in **#1800**. Read viewer data directly from the homeserver so the marker mechanism is no longer needed and multi-device drift is solved by construction.

---

<details>
<summary>~~Earlier plans considered (collapsed for reference)~~</summary>

### Earlier (now superseded): IndexedDB-backed marker

We started building a persistent marker as a new IndexedDB table `viewer_tag_mutations` with primary key `id = "${postId}:${label}:${taggerId}"` and indexes on `postId` and `expiresAt`. This required a DB schema change and a `NEXT_PUBLIC_DB_VERSION` bump (3 → 4). The protection profile was the same as the chosen sessionStorage approach, but with a higher schema cost. Replaced with sessionStorage because:

- sessionStorage gives the same race + drift + F5 protection.
- No schema migration, no DB version bump.
- Less code, smaller test surface.
- IndexedDB's only added feature was cross-tab marker sharing, which is a low-value edge case.

### Earlier (now superseded): in-memory Map only

Brief consideration of using a module-level `Map<string, Marker>` with no persistence. Rejected because the marker would be lost on F5 refresh during the protection window — leaving the visible flicker bug partially unfixed for that case.

### Earlier (now superseded): SSE-based sync (PR #1787 style)

Considered subscribing to the homeserver event stream for tag URIs, like the mute list does in #1787. Rejected for short-term because:

- A tag-event coordinator would need to maintain a global cache of viewer tags across all posts, including posts the user only interacted with once. That's a lot of data of low ongoing value.
- Tag events come keyed by tag URI hash, so we'd still need to fetch each tag JSON and resolve the post URI to use it.
- The tag UX does not need real-time push — refreshing on the post page is fine.

This belongs in the long-term plan (#1800), not the short-term fix.

#### Context

Follow-up to #1721. The UI fix in #1721 (passing `viewer_id` to `getPostTags`) made `relationship` correct on initial fetch. Once we started using the app with that fix, a separate eventual-consistency race surfaced.

#### The race

`usePostTags.fetchInitialTags` runs an in-flight Nexus fetch on every mount. While that fetch is in-flight, the user can perform a local-first tag commit. When the (now-stale) response comes back, `LocalPostTagService.mergeTags` writes it to IndexedDB, overwriting the just-mutated local state.

Concrete repro on `/home`:

1. Post is rendered with a tag the viewer has previously tagged (highlighted).
2. Viewer clicks the tag to remove it.
3. `commitDelete` runs, IndexedDB is updated, `useLiveQuery` refires, UI shows highlight off — correct.
4. ~1 second later the in-flight Nexus response from the page-mount fetch returns. Nexus has not yet indexed the homeserver delete, so it replies with the old state (`relationship: true`, viewer in `taggers`). `mergeTags` writes that into IndexedDB.
5. UI shows the tag highlighted again.

Same shape on the create path: tag click → highlights → mid-flight fetch returns stale `relationship: false` → `mergeTags` overwrites → highlight disappears.

The /home → /post[id] navigation case (#1721 follow-up reported earlier) has the same root cause but only manifests on remount.

#### Why the simple TTL gate is not enough

A first attempt added a TTL freshness gate inside `usePostTags.fetchInitialTags`:

```ts
const stalePostIds = await TtlController.findStalePostsByIds({
  postIds: [postId],
  ttlMs: TTL_POST_MS,
});
if (stalePostIds.length === 0) return; // skip fetch
```

This works for the **navigation** case (commit bumps `post_ttl.lastUpdatedAt`, next mount sees fresh, skips) but does **not** address the same-mount race: by the time the click happens, `fetchInitialTags` has already issued the Nexus request. The gate only protects fetch *initiation*, not in-flight responses.

Also, `post_ttl.lastUpdatedAt` is bumped by both viewer commits and stream / TTL-coordinator refreshes, so any defensive check based on that timestamp can't distinguish "viewer mutated" from "stream refetched".

#### Proposed solutions (pick one)

**A. Defensive timestamp check in `PostApplication.fetchTags`.** Capture a viewer-mutation marker before issuing the Nexus call; after the response, if the marker has advanced, skip the merge. Simple, targeted, but needs a viewer-specific marker (post_ttl alone has false positives).

**B. Viewer-mutation marker (in-memory `Map<postId, ts>` or separate IndexedDB table).** Updated only by `LocalPostTagService.create/delete`. `mergeTags` (or the fetch caller) reads it to skip writes when a recent viewer mutation overlaps the in-flight fetch. Cleanest separation of concerns; new mechanism.

**C. AbortController for in-flight tag fetches.** When `commitCreate`/`commitDelete` runs for a postId, abort any in-flight `fetchInitialTags` for that same postId. Most architecturally correct, biggest change (controller signatures + hook plumbing).

**D. Preserve viewer-owned fields inside `mergeTags`.** `relationship` and the viewer's presence in `taggers` are owned by `addTagger`/`removeTagger`; the merge only updates other-user taggers and counts. Discussed and previously rejected because: (1) makes `mergeTags` quietly diverge from "server is source of truth", (2) `bulkSave` (used by stream / TTL refresh) would still bypass it, (3) misses sole-tagger-delete and multi-device sync edge cases.

#### Snippet of the deferred work (TTL gate, kept here for reference)

```ts
// src/hooks/usePostTags/usePostTags.tsx — inside fetchInitialTags
const stalePostIds = await TtlController.findStalePostsByIds({
  postIds: [postId],
  ttlMs: TTL_POST_MS,
});
if (stalePostIds.length === 0) {
  if (!stale) setHasFetched(true);
  return;
}
```

Helps the navigation case but does not solve the same-mount race; full fix needs one of A–D above.

</details>

### PR DESCRIPTION
**Pull request by Taewa**
## Summary
  Fixes a tagging bug where a user's freshly added or removed tag could "snap back" after a Nexus refresh,
  because Nexus's indexer is several minutes behind the homeserver write. Closes #1781.

  ## Root cause
  - After a user toggles a tag, the next Nexus response for that post is often still serving pre-toggle stale data
   (indexer lag, ~minutes).
  - The old merge step blindly took the union of local + Nexus and used `max(local, nexus)` for the count,
  so a stale Nexus response could resurrect a just-deleted tag or undercount a just-added one.
  - Separately, a delete that returned 404 from the homeserver was treated as a failure and rolled back
  locally, leaving the user with a "ghost" tag they couldn't remove (the homeserver kept 404-ing on every
  retry).

  ## Fix
  - **Viewer-mutation marker** — when the viewer tags or untags a post, drop a short-lived (~5 min) marker
  in sessionStorage; the merge step trusts the marker over Nexus for that one viewer slot.
  - **Per-field merge policy** — count comes from Nexus (with ±1 correction when a marker disagrees), the
  viewer's slot in the tagger list follows the marker, other taggers stay as a union (Nexus only ships a
  truncated top-N), labels not in the Nexus response are left alone (they may be on a later page).
  - **Treat homeserver 404 on delete as success** — the tag is already gone, no rollback, no ghost.
  - **Logout sweep** — clear the current user's markers before the next user signs in on the same tab.

  ## Notes
  - Scope is post tags only — user tags and other entity tags are unchanged.
  - Marker TTL (`MARKER_TTL_MS`, 5 min) is a tunable in `src/config/viewerTagMarker.ts`; pick the indexer
  lag ceiling, not a "freshness" budget.
  - Markers live in sessionStorage, so they're per-tab and naturally gone on tab close; the logout sweep
  covers same-tab user switches.
  
  ## Result
  
### 1.

https://github.com/user-attachments/assets/87215b95-d99b-4e3c-9c5a-2b8f9e1267df

### 2.

https://github.com/user-attachments/assets/53fc7af9-8819-45b2-b81d-36f43704e3f1

### CODE REVIEW

**Comment by Taewa** [comment]
@greptile-apps
Re: the load-bearing invariant — I considered a `mutateAndMark(op, fn)` wrapper but skipped it: only two call sites, and `create` / `delete` differ enough (idempotent early-return vs not) that a shared wrapper would hurt readability. 
A missing marker would also break the `mergeTags` tests immediately, so the risk is bounded. Happy to revisit if a third mutation path appears.

**Comment by Taewa** [comment]
@greptile-apps
I also updated `sweepExpired` related code. Review this PR again plz

File: src/core/services/local/tag/post/viewerTagMarkerStorage.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+function buildKey(pubky: Pubky, postId: string, label: string): string {
+  return `${VIEWER_TAG_MARKER_STORAGE_PREFIX}${pubky}:${postId}:${label.toLowerCase()}`;
+}
+
+function userPrefix(pubky: Pubky): string {
+  return `${VIEWER_TAG_MARKER_STORAGE_PREFIX}${pubky}:`;
+}
+
+function isStorageAvailable(): boolean {
```
Comment (infin1t3) [inline src/core/services/local/tag/post/viewerTagMarkerStorage.ts]: This is a Next.js app but we do not use any of the SSR functionalities, given this is an offline first app. So `window` and storage should always be available and I am not sure we need this check.

File: src/core/services/local/tag/post/viewerTagMarkerStorage.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  return typeof window !== 'undefined' && typeof window.sessionStorage !== 'undefined';
+}
+
+function safeRemove(key: string): void {
```
Comment (infin1t3) [inline src/core/services/local/tag/post/viewerTagMarkerStorage.ts]: Not sure we need this wrapper function, instead of calling `window.sessionStorage.removeItem(key);` directly.

Especially since a bit down we use other similar methods without a wrapper - `window.sessionStorage.getItem(key);`

And even if we do want to have some wrapper for storages, may be better to make that into a separate util.

File: src/core/services/local/tag/post/tag.post.ts
Comment (infin1t3) [inline src/core/services/local/tag/post/tag.post.ts:35]: @returns {boolean} true if local state changed; false if the tagger already had this tag (idempotent — no writes, no viewer marker)

**Review by infin1t3 — CHANGES_REQUESTED** [review]

File: src/core/services/local/tag/post/tag.post.ts
Comment (Taewa) [reply src/core/services/local/tag/post/tag.post.ts:35]: Done

File: src/core/services/local/tag/post/viewerTagMarkerStorage.ts (hunk 1 above)
Comment (Taewa) [reply src/core/services/local/tag/post/viewerTagMarkerStorage.ts]: Done

File: src/core/services/local/tag/post/viewerTagMarkerStorage.ts (hunk 2 above)
Comment (Taewa) [reply src/core/services/local/tag/post/viewerTagMarkerStorage.ts]: Done

**Review by infin1t3 — APPROVED** [review]
Tested at - `8f0be7ee85183f46d8fe276a5103ca84eb2f9cab` - LGTM. Approved.
https://github.com/user-attachments/assets/e9604c87-9990-41c7-b3ec-8d0c99d0e65b
